### PR TITLE
Mutex code cleanup and removal of unused types

### PR DESCRIPTION
**Pull request by ViktorHofer**
@jkotas should I also align Interop code and move it into shared?

### CODE REVIEW

**Comment by ViktorHofer** [comment]
I just noticed that most of the cleanup was already done in corert.

File: src/mscorlib/src/System/Threading/Mutex.cs [hunk 1]
```diff
@@ -264,20 +151,83 @@ private static OpenExistingResult OpenExistingWorker(string name, MutexRights ri
...
         {
-            if (Win32Native.ReleaseMutex(safeWaitHandle))
+            if (!Win32Native.ReleaseMutex(safeWaitHandle))
             {
+                throw new ApplicationException(SR.Arg_SynchronizationLockException);
             }
-            else
+        }
+
+        internal class MutexTryCodeHelper
```
Comment (jkotas) [inline src/mscorlib/src/System/Threading/Mutex.cs]: If you would like, you can get rid of this too and just replace it with try/finally block

**Comment by jkotas** [comment]
> should I also align Interop code and move it into shared?
> I just noticed that most of the cleanup was already done in corert.

Ideally, both of these should be done together. I would wait with this - the synchronization/threading is one of the harder areas to unify.

**Comment by jkotas** [comment]
LGTM otherwise

**Comment by ViktorHofer** [comment]
PTAL. I aligned the code more to the corert version to simplify converging later.

File: src/mscorlib/src/System/Threading/Mutex.cs [hunk 2]
```diff
@@ -2,218 +2,123 @@
...
-                // Empty name is treated as an unnamed mutex. Set to null, and we will check for null from now on.
-                name = null;
-            }
-#if PLATFORM_WINDOWS
-            if (name != null && System.IO.Path.MaxPath < name.Length)
-            {
-                throw new ArgumentException(SR.Format(SR.Argument_WaitHandleNameTooLong, Path.MaxPath), nameof(name));
-            }
-#endif // PLATFORM_WINDOWS
+            VerifyNameForCreate(name);
```
Comment (jkotas) [inline src/mscorlib/src/System/Threading/Mutex.cs]: This used to be under Windows ifdef, but it is enabled for all platforms now

File: src/mscorlib/src/System/Threading/Mutex.cs [hunk 3]
```diff
@@ -224,16 +129,16 @@ private static OpenExistingResult OpenExistingWorker(string name, MutexRights ri
             {
                 throw new ArgumentException(SR.Argument_EmptyName, nameof(name));
             }
+
 #if !PLATFORM_UNIX
-            if (System.IO.Path.MaxPath < name.Length)
+            if (Path.MaxPath < name.Length)
```
Comment (jkotas) [inline src/mscorlib/src/System/Threading/Mutex.cs]: Should this use the `VerifyNameForCreate` method?

File: src/mscorlib/src/System/Threading/Mutex.cs [hunk 4]
```diff
@@ -2,218 +2,123 @@
...
+                throw Win32Marshal.GetExceptionForWin32Error(errorCode, name);
             }
-        }
 
-        public static bool TryOpenExisting(string name, out Mutex result)
-        {
-            return OpenExistingWorker(name, (MutexRights)0, out result) == OpenExistingResult.Success;
+            createdNew = errorCode != Interop.Errors.ERROR_ALREADY_EXISTS;
+            SafeWaitHandle = mutexHandle;
+            hasThreadAffinity = true;
```
Comment (jkotas) [inline src/mscorlib/src/System/Threading/Mutex.cs]: It looks like you have deleted this in other case, but not this one.

File: src/mscorlib/src/System/Threading/Mutex.cs (hunk 3 above)
Comment (ViktorHofer) [reply src/mscorlib/src/System/Threading/Mutex.cs]: The `VerifyNameForCreate` method also checks if name is not null which we don't need in this case but I guess it doesn't really matter performance-wise. I will reuse it here

File: src/mscorlib/src/System/Threading/Mutex.cs (hunk 4 above)
Comment (ViktorHofer) [reply src/mscorlib/src/System/Threading/Mutex.cs]: Do you think it's safe to remove them everywhere now that I'm no longer using the `RuntimeHelpers.CleanupCode` method?

File: src/mscorlib/src/System/Threading/Mutex.cs (hunk 2 above)
Comment (ViktorHofer) [reply src/mscorlib/src/System/Threading/Mutex.cs]: On corert it's throwing when a name is provided: https://github.com/dotnet/corert/blob/master/src/System.Private.CoreLib/src/System/Threading/Mutex.Unix.cs#L16

Should this be the same here? I'm not sure about this...

Comment (jkotas) [reply src/mscorlib/src/System/Threading/Mutex.cs]: Expected - CoreRT does not support named mutexes on Unix today.

File: src/mscorlib/src/System/Threading/Mutex.cs (hunk 4 above)
Comment (jkotas) [reply src/mscorlib/src/System/Threading/Mutex.cs]: It think so. It is also in WaitHandle and a few FCalls - it can be removed there as well.

**Comment by ViktorHofer** [comment]
@jkotas should we merge that for now and I will look into other classes like SafeHandle in a separate PR?

File: src/mscorlib/src/System/Threading/Mutex.cs [hunk 5]
```diff
@@ -2,216 +2,123 @@
...
-            switch (OpenExistingWorker(name, rights, out result))
+            if (mutexHandle.IsInvalid)
             {
-                case OpenExistingResult.NameNotFound:
-                    throw new WaitHandleCannotBeOpenedException();
-
-                case OpenExistingResult.NameInvalid:
+                mutexHandle.SetHandleAsInvalid();
+#if PLATFORM_UNIX
+                if (errorCode == Win32Native.ERROR_FILENAME_EXCED_RANGE)
```
Comment (jkotas) [inline src/mscorlib/src/System/Threading/Mutex.cs]: Nit: Should this rather be `Interop.Errors.ERROR_FILENAME_EXCED_RANGE`?

**Comment by jkotas** [comment]
@dotnet-bot test Windows_NT x64 corefx_baseline
@dotnet-bot test Ubuntu x64 corefx_baseline

**Comment by jkotas** [comment]
> should we merge that for now 

ok. (once the corefx legs are green)

**Comment by ViktorHofer** [comment]
the output from the corefx legs is unfortunately unusable :(

**Comment by jkotas** [comment]
The corefx legs are failing with: `'System.Runtime.Intrinsics.X86.Avx.Insert(System.Runtime.Intrinsics.Vector256, System.Single*, System.Byte)' does not exist in the implementation`

@fiigii Was it intentional that you have deleted `byte index` from several Avx.Insert overloads in your change #14164? If not, could you please submit PR to fix it?

**Comment by jkotas** [comment]
@ViktorHofer #14231 should fix the corefx break. We should retry after it goes in.

**Comment by ViktorHofer** [comment]
@dotnet-bot test this please

**Comment by ViktorHofer** [comment]
@dotnet-bot test Windows_NT x64 corefx_baseline
@dotnet-bot test Ubuntu x64 corefx_baseline

**Comment by ViktorHofer** [comment]
Failing legs are unrelated to my changes... Some FileSystemWatcher tests are failing

**Comment by ViktorHofer** [comment]
@dotnet-bot test Ubuntu x64 corefx_baseline
@dotnet-bot test Ubuntu x64 Formatting

**Comment by ViktorHofer** [comment]
Maybe the errors are related. I need to check...

```
System.IO.Tests.FileSystemWatcherTests_netstandard17.EndInit_ResumesPausedEnableRaisingEvents
System.IO.Tests.FileSystemWatcherTests_netstandard17.EndInit_ResumesPausedEnableRaisingEvents(setBeforeBeginInit: True)
System.IO.Tests.FileSystemWatcherTests_netstandard17.EndInit_ResumesPausedEnableRaisingEvents(setBeforeBeginInit: False)
System.IO.Tests.File_Create_Tests.FileSystemWatcher_File_Create
```

**Comment by jkotas** [comment]
The FileSystemWatcher failures are #14154

**Comment by ViktorHofer** [comment]
@dotnet-bot test Ubuntu x64 corefx_baseline

**Comment by ViktorHofer** [comment]
@dotnet-bot test Ubuntu x64 Checked Build and Test (Jit - CoreFx)

**Pull request merged**